filecoin-proofs/src/api/seal.rs

```diff
@@ -556,6 +560,92 @@ pub fn seal_commit_phase2<Tree: 'static + MerkleTreeTrait>(
     Ok(out)
 }
 
+/// Returns the path to the sealing key (i.e. last replication layer) stored on disk.
+#[inline]
+pub fn sealing_key_path<Tree: MerkleTreeTrait>(
+    pc1_output: &SealPreCommitPhase1Output<Tree>,
+) -> anyhow::Result<PathBuf> {
+    pc1_output
+        .labels
+        .labels
+        .last()
+        .map(|layer_store| StoreConfig::data_path(&layer_store.path, &layer_store.id))
+        .with_context(|| "PC1 output's label store is empty")
+}
+
+/// Returns a range of nodes read from the sealing key (i.e. last replication layer) stored on disk.
+#[inline]
+pub fn read_sealing_key_range<Tree: MerkleTreeTrait>(
+    pc1_output: &SealPreCommitPhase1Output<Tree>,
+    node_range: Range<usize>,
+) -> anyhow::Result<Vec<<Tree::Hasher as Hasher>::Domain>> {
+    pc1_output
+        .labels
+        .labels_for_last_layer()?
+        .read_range(node_range)
+        .with_context(|| "failed to read sealing key range")
+}
+
+/// Returns a range of nodes from unreplicated data stored on disk.
+#[inline]
+pub fn read_sealing_data_range<Tree: MerkleTreeTrait>(
+    pc1_output: &SealPreCommitPhase1Output<Tree>,
+    node_range: Range<usize>,
+) -> anyhow::Result<Vec<DefaultPieceDomain>> {
+    read_nodes_from_file(
+        StoreConfig::data_path(&pc1_output.config.path, &pc1_output.config.id).as_path(),
+        node_range,
+    )
+}
+
+/// Returns a range of nodes from a replica stored on disk.
+#[inline]
+pub fn read_replica_range<Tree: MerkleTreeTrait>(
+    replica_path: &Path,
+    node_range: Range<usize>,
+) -> Result<Vec<<Tree::Hasher as Hasher>::Domain>> {
+    read_nodes_from_file(replica_path, node_range)
+}
+
+/// Encodes a range of unreplicated data nodes (stored on disk) into replica nodes using the last
+/// sealing layer as the encoding key.
+pub fn encode_data_range<Tree: MerkleTreeTrait>(
+    pc1_output: &SealPreCommitPhase1Output<Tree>,
+    node_range: Range<usize>,
+) -> Result<Vec<<Tree::Hasher as Hasher>::Domain>> {
+    let data_nodes = read_sealing_data_range(pc1_output, node_range.clone())?;
+    let key = read_sealing_key_range(pc1_output, node_range)?;
+    data_nodes
+        .into_iter()
+        .zip(key)
+        .map(|(data_node, key)| {
+            <Tree::Hasher as Hasher>::Domain::try_from_bytes(data_node.as_ref())
+                .with_context(|| "failed to convert TreeD leaf bytes into TreeR domain")
+                .map(|data_node| storage_proofs_porep::encode(key, data_node))
+        })
+        .collect()
+}
+
+/// Decodes a range of replica nodes (stored on disk) into unreplicated data nodes using the last
+/// sealing layer as the encoding key.
+pub fn decode_replica_range<Tree: MerkleTreeTrait>(
+    pc1_output: &SealPreCommitPhase1Output<Tree>,
+    replica_path: &Path,
```

Comment:
I'd prefer if encode/decode just operated on byte streams. Lotus / other programs can make sure that those byte streams are properly aligned at node boundaries - those APIs should work with pipe FDs.

This is needed for cases where the last sealing layer, and replica are on different computers, and we want to decode bits of data on the fly

Reply:
It would also be preferable if `pc1_output: &SealPreCommitPhase1Output` was specified as an FD / Path to the last layer / sectorkey